Extending wxFont API & OSX Core Text Implementation

csomor opened the pull request:
Hi

I've committed the first stage of changing, buildable on msw and osx.

- extending the font weight API: having more wxFontWeight constants and exposing an integer via `SetNumericWeight` / `GetNumericWeight`

- allowing fractional point sizes: `SetPointSize` gets a float argument and `GetFractionalPointSize` as an additional Getter to avoid breaking changes on GetPointSize which still returns an integer value.

Thanks for thought and opinions

Stefan

csomor committed:
Switch to pure Core Text Implementation, Start extended Font API

csomor committed:
mac fixes

csomor committed:
First msw implementation

csomor committed:
Fixing paste error

csomor committed:
fixing typo

csomor committed:
Rearranging lines to former fallthrough order

csomor committed:
Blind fixes for covering new abstract methods

csomor committed:
Blind gtk implementations

csomor committed:
Fixing according to travis ..

csomor committed:
Removing method defined in base

catalinr commented on src/unix/fontutil.cpp:
```diff
@@ -252,20 +257,28 @@ void wxNativeFontInfo::SetStyle(wxFontStyle style)
 {
-    switch (weight)
+    // deal with compatibility constants
+    if (weight >= 90 && weight <= 92)
```
some more `if ( ... )` spaces

catalinr reviewed (commented):
Knitpicks about spaces.

csomor commented on src/common/fontcmn.cpp:
```diff
@@ -988,21 +1037,66 @@ bool wxNativeFontInfo::FromUserString(const wxString& s)
             weightfound = true;
         }
-        else if ( token == wxT("bold") || token == _("bold") )
+        else if (token == wxT("normal") || token == _("normal"))
```
I don't see the problem here actually, clang-format keeps it that way

catalinr replied:
The old line that is shown in this partial diff, and the unchanged line 1050 above, have spaces before and after the `if` condition, or IOW 1 space after `(` and 1 space before `)`. Automatic formatting may change that to something more common though.

csomor replied:
ok, I see that our own style guide has the spaces there, I'll check my clang-format.config

csomor committed:
formatting adaptions

csomor commented:
Thanks for looking over the formatting. I haven't found the option to have spaces after keyword expressions but not around simple method-calls-parentheses. Does anybody have a suggestion on tool support for our kind of spacing rules ?

vadz reviewed (commented):
Sorry, I'm too tired to review this in details now, but as I'm leaving tomorrow I wanted to have at least a quick look.

Globally it looks fine, although I don't quite understand why do we need a cache in wxMac now and whether we'll need it in the other ports.

If I could make one small request for the future, could you please separate whitespace-only changes into their own commits (or even PRs)? When they're mixed up with the significant changes, it makes them more difficult to review. TIA!

csomor commented:
Thanks at lot for your first sight.
The **caching** was already there before, because at least on some macOS versions this really became a bottleneck and hampered performance. So no, I wouldn't introduce this on OSes which didn't have it before.

csomor commented on src/osx/carbon/font.cpp:
```diff
@@ -180,15 +194,89 @@ wxFontRefData::wxFontRefData(const wxFontRefData& data) : wxGDIRefData()
 {
-    m_fontValid = false;
+    // free additional allocated resources first
+    Free();
```
True, Thanks

csomor commented on src/osx/carbon/font.cpp:
```diff
@@ -199,189 +287,157 @@ wxFontRefData::~wxFontRefData()
-        if (m_info.m_style == wxFONTSTYLE_ITALIC || m_info.m_style == wxFONTSTYLE_SLANT)
-            traits |= kCTFontItalicTrait;
+    // use font caching, we cache a font with a certain size and a font with just any size for faster creation
+    wxString lookupnameNoSize = wxString::Format("%s_%d_%d", m_info.GetFamilyName(), (int)m_info.GetStyle(), (int)m_info.GetNumericWeight());
```
no, your right

csomor commented on src/xrc/xmlres.cpp:
```diff
@@ -2306,10 +2306,24 @@ wxFont wxXmlResourceHandlerImpl::GetFont(const wxString& param, wxWindow* parent
         wxString weight = GetParamValue(wxT("weight"));
-        if (weight == wxT("bold"))
-            iweight = wxFONTWEIGHT_BOLD;
+        if (weight == wxT("thin"))
```
I'll do that

csomor commented on src/xrc/xmlres.cpp:
```diff
@@ -2306,10 +2306,24 @@ wxFont wxXmlResourceHandlerImpl::GetFont(const wxString& param, wxWindow* parent
+            iweight = wxFONTWEIGHT_HEAVY;
+        else if (weight == wxT("extraheavy"))
+            iweight = wxFONTWEIGHT_EXTRAHEAVY;
         else if (weight != wxT("normal"))
```
I'll leave that, I'm not fluent with relax-ng

csomor committed:
Extending the schema definition for new weights

csomor committed:
fixing typo, using wxRound, other fixes according to comments

csomor commented on src/msw/font.cpp:
```diff
@@ -461,13 +478,19 @@ wxFontStyle wxNativeFontInfo::GetStyle() const
 {
-    if ( lf.lfWeight <= 300 )
-        return wxFONTWEIGHT_LIGHT;
+    int weight = ((GetNumericWeight() + 50) / 100) * 100;
```
thanks, I'll do so

csomor commented on src/common/fontcmn.cpp:
```diff
@@ -816,13 +836,41 @@ wxString wxNativeFontInfo::ToUserString() const
+            break;
+
+        case wxFONTWEIGHT_EXTRALIGHT:
+            desc << _(" extralight");
```
ok, I'll go for the "extra light"

csomor commented on include/wx/osx/font.h:
```diff
@@ -99,26 +100,28 @@ class WXDLLIMPEXP_CORE wxFont : public wxFontBase
+    virtual wxFontEncoding GetEncoding() const wxOVERRIDE;
+    virtual const wxNativeFontInfo *GetNativeFontInfo() const wxOVERRIDE;
 
     virtual bool IsFixedWidth() const;
```
true

vadz commented on src/gtk/font.cpp:
```diff
@@ -373,6 +379,13 @@ wxFontWeight wxFont::GetWeight() const
 
+int wxFont::GetNumericWeight() const
+{
+    wxCHECK_MSG( IsOk(), wxFONTWEIGHT_MAX, wxT("invalid font") );
```
`wxS()` is slightly more efficient (no conversion needed), so if you care about this, it can be used, but I don't think it's really a big concern for the assert messages (we definitely don't care about speed of asserting, so it's just a few extra bytes in the library size) and using nothing is simpler and more clear.

csomor replied:
thanks for the clarification

csomor committed:
changes according to suggestions

csomor committed:
fixing init order, before the init of m_info was overridden by Init()

csomor committed:
cleanup, dealing with wxDEFAULT only in wxFont int-size constructor

csomor committed:
redo

csomor committed:
redo

csomor committed:
redo

vadz commented on src/osx/carbon/font.cpp:
```diff
@@ -909,42 +1041,87 @@ wxString wxNativeFontInfo::ToString() const
 wxString wxNativeFontInfo::GetFaceName() const
 {
-    return m_faceName;
+#if 0
```
Could you use `wxLogTrace("ctfont", ...)` instead? This could stay enabled in all builds unless this function is very performance-critical, but I don't think it is (and if it is, let's put this inside `#if wxDEBUG_LEVEL >= 2` which is still better than `#if 0`).

csomor replied:
committed

vadz commented on src/osx/carbon/font.cpp:
```diff
@@ -1035,21 +1235,174 @@ void wxNativeFontInfo::SetStrikethrough(bool strikethrough)
     }
 }
 
+// code that still may be useful
```
I'd really prefer to keep it in a side branch instead of here, it's confusing to see commented out code like this and it appears in search results etc.

csomor replied:
cleanup committed

vadz commented on src/unix/fontutil.cpp:
```diff
@@ -252,20 +258,28 @@ void wxNativeFontInfo::SetStyle(wxFontStyle style)
 void wxNativeFontInfo::SetWeight(wxFontWeight weight)
 {
-    switch (weight)
+    // deal with compatibility constants
```
Would be nice to provide a single function doing this instead of repeating this (some day to be removed) code in multiple files. If we had a function, it would be easier to drop it, and all references to it, later.

csomor replied:
yes, agreed, I'll put something in fontcmn.cpp

csomor replied:
ok, done that, thanks for the feedback

vadz reviewed (approved):
Sorry, I completely forgot that I had promised to review this more carefully. I just reread everything again and I don't have any important remarks left (just a couple of minor ones...), thanks a lot.

The only thing that prevents this from being merged are the updates to the documentation: we need to update `interface/wx/font.h` (using `@since 3.1.2` for the new functionality) and `docs/changes.txt` to explain the changes, especially backwards-incompatible ones if any (I'm not even sure if there are?).

Other than that, let's merge this a.s.a.p. If you'd like, I could even do the docs update and the merge myself, please let me know.

csomor committed:
Cleanup

Removing obsolete code snippets, proper traces for font names

csomor committed:
Moving common code

Only the Get/SetNumericWeight calls should now be implemented in the native part, the ‚old‘ Get/SetWeight are common code and use the numeric counterparts.

csomor committed:
Updating docs

csomor commented:
I've added the docs now as well, could you please glance things over ? I'd merge it in afterwards, Thanks

vadz reviewed (commented):
No real comments, just a few tiny remarks about the docs.

Also, we still need to mention this in the changelog (`docs/changes.txt`).

Thanks again!

csomor committed:
commit wa missing changes.txt

csomor committed:
Doc fixes

vadz commented on docs/changes.txt:
```diff
@@ -483,6 +483,8 @@ All (GUI):
 - Allow suppressing warnings from wxImage::LoadFile().
 - Allow customizing wxRibbon highlight colours (wxBen).
 - Enable selecting opacity in generic wxColourPicker.
+- Extend the support for font weights to a bigger range
```
You're going to hate me for nitpicking, but could you please add full stops at the end of the new lines here for consistency with all the others?

Other than that, I don't see anything, let's merge this!

csomor replied:
No, I really appreciate your comments. It helps getting better results, and I'll try to remember :-)

csomor committed:
Full stops added

vadz commented:
This is annoying. It looks like we may need `SetFractionalPointSize()`, finally... Alternatively, we could change `float` to `double` here as conversion from `int` to `double` should be lossless and not result in such warnings. Finally, we could also make `SetPointSize()` non-virtual and provide 2 overloads of it, but this might break the existing code which uses `SetPointSize(long)`, for example, as there will now be an ambiguous conversion from `long` to either `int` or `double`.

So the most compatible fix is probably to switch to `double`, what do you think?

csomor commented:
I'm just seeing either the symmetry of Get/SetFractionalPointSize or the double param. Current code would not use a fractional point size parameter, so while changing the type using a different API would seem to be ok IMHO, switching from float to double - if not resulting in any warnings - be preferable for me because it's one function less in the API...

MaartenBent commented:
Using `double` fixes the warning.

csomor commented:
I guess we should change the GetFractionalPointSize return type to double as well ?

vadz commented:
For me it's basically the question of whether we think it's useful to be explicit about using non-integer point sizes (e.g. because some platforms might not support it?) or if, on the contrary, we want to hide the difference between them.

To summarize, we have 2 acceptable choices:

1. Add `SetFractionalPointSize(float)`, restore `SetPointSize(int)` which will just forward to it.
2. Change `SetPointSize()` to take `double` and change `GetFractionalPointSize()` to return `double`.

The first one is 100% backwards compatible, but the second one is 99% compatible too, so from this point of view both should be fine. The main problem with the latter solution I see is the asymmetry between setter (one only) and getters (two). But I also agree that having one less setter could be seen as an advantage.

Personally I'd probably do (1), but I'm really fine with both and if you (Stefan) prefer (2), we can also do it like this -- your choice, really.

BTW, if you'd like to do (1), I could do it while fixing wxX11 build as this would require changing the same files and redoing all the same builds anyhow. Please let me know.

csomor commented:
would you think that in a future version we would ever change the return type of GetPointSize ? I've already prepared the files for (2) but in the end I guess having 100 % backward compatibility means less questions asked and more time for coding ;-) and if the return type of GetPointSize() would be int forever, the asymmetry would probably always hurt me ... so then I'd favour (1)

csomor commented:
ok, let's do (1) and tell me whether I should do it

vadz commented:
I guess I'll do this, together with the X11 fix.

csomor commented:
thanks
